engine/src/Interpreter/interpreter_ops.cuh

```diff
@@ -677,253 +391,358 @@ private:
+					case cudf::type_id::TIMESTAMP_NANOSECONDS:
+						computed = cudf::timestamp_ns{static_cast<cudf::timestamp_ns::rep>(left_value)};
+						break;
+					case cudf::type_id::TIMESTAMP_DAYS:
+						computed = cudf::timestamp_ns{cudf::timestamp_D{static_cast<cudf::timestamp_D::rep>(left_value)}};
+						break;
+					case cudf::type_id::TIMESTAMP_SECONDS:
+						computed = cudf::timestamp_ns{cudf::timestamp_s{static_cast<cudf::timestamp_s::rep>(left_value)}};
+						break;
+					case cudf::type_id::TIMESTAMP_MILLISECONDS:
+						computed = cudf::timestamp_ns{cudf::timestamp_ms{static_cast<cudf::timestamp_ms::rep>(left_value)}};
+						break;
+					case cudf::type_id::TIMESTAMP_MICROSECONDS:
+						computed = cudf::timestamp_ns{cudf::timestamp_us{static_cast<cudf::timestamp_us::rep>(left_value)}};
+						break;
+					default:
+						// should not reach here, invalid conversion
+						assert(false);
+						break;
+					}
+					store_data_in_buffer(static_cast<int64_t>(computed.time_since_epoch().count()), buffer, output_position);
 				}
 			} else if(oper == BLZ_IS_NULL) {
-				computed = !left_valid;
+				store_data_in_buffer(static_cast<int64_t>(!left_valid), buffer, output_position);
 			} else if(oper == BLZ_IS_NOT_NULL) {
-				computed = left_valid;
+				store_data_in_buffer(static_cast<int64_t>(left_valid), buffer, output_position);
 			}
 
-			if(oper == BLZ_IS_NULL || oper == BLZ_IS_NOT_NULL) {
-				setColumnValid(row_valids, output_position, true);
-			} else {
-				setColumnValid(row_valids, output_position, left_valid);
-			}
-			store_data_in_buffer<OutputTypeOperator>(computed, buffer, output_position);
+			bool out_valid = (oper == BLZ_IS_NULL || oper == BLZ_IS_NOT_NULL) ? true : left_valid;
+			setColumnValid(row_valids, output_position, out_valid);
 		}
 	}
 
 private:
-	cudf::size_type num_columns;
-	cudf::size_type num_final_outputs;
-
-	void ** column_data;  // these are device side pointers to the device pointer found in gdf_column.data
-	void ** output_data;
-	temp_gdf_valid_type ** valid_ptrs;
-	temp_gdf_valid_type ** valid_ptrs_out;
-
-	cudf::type_id * input_column_types;
-	column_index_type * left_input_positions;
-	column_index_type * right_input_positions;
-	column_index_type * output_positions;
-	column_index_type * final_output_positions;  // should be same size as output_data, e.g. num_outputs
-
-	int num_operations;
-	cudf::type_id * input_types_left;
-	cudf::type_id * input_types_right;
-	cudf::type_id * output_types;
-	cudf::type_id * final_output_types;
-	gdf_binary_operator_exp * binary_operations;
-	gdf_unary_operator * unary_operations;
-	int64_t * scalars_left;  // if these scalars are not of invalid type we use them instead of input positions
-	int64_t * scalars_right;
-
-	cudf::size_type * null_counts_inputs;
-	cudf::size_type * null_counts_outputs;
-
-	int64_t * temp_valids_in_buffer;
-	int64_t * temp_valids_out_buffer;
-
-	int threadBlockSize;
+	cudf::mutable_table_device_view out_table;
```

Comment:
How are you ensuring that this information is available to the gpu?

Reply:
How are you making sure that this is all in one contiguous buffer?

Reply:
Not sure what you mean,  this class is meant to be used in kernels as it's name implies

Reply:
ok i guess we arent making sure of that anymore. We need to do some performance testing to see if there is a cost to this. Because the plan was all in one contiguous buffer before you knew that every thread was basically reading from cache when looking at the plan.